keps/sig-node/20200303-shared-pod-limits.md

```diff
@@ -0,0 +1,432 @@
+   |   +-- container1 (first container)
+   |   |
+   |  = ... =
+   |   |
+   |   +-- containerN (N-th container)
+   |
+   +-- QoS CGroup (one of burstable, or besteffort)
+       |
+       \ pod 
+          |
+          +-- container0 (pause container)
+          |
+          +-- container1 (first container)
+          |
+        = ... =
+          |
+          +-- containerN (N-th container)
+</pre>
+
+The proposal in this KEP is to allow users to define resources on the Pod level itself. Both resource requests and limits are allowed. The scheduler
+algorithm will consider the pod-level resources if they exist in preference to the container-level resources. If no pod-level resource section is provided,
+then the scheduler will continue to operate as it does in the current implementation. The QoS definitions will be updated to consider the pod-level resources
+when classifying a pod into one of the QoS levels. If the pod level doesn't include any resource limits, then kubelet will continue to function as it does today.
+
+The current implementation will set the cgroup limits for the memory/CPU resources on the pod level of the hierarchy only in the following cases:
+1. The QoS level is `Guaranteed`. 
+1. The QoS level is `Burstable` and all of the containers specify a limit for the relevant resource. 
+
+In the current implementation pods are considered `Guaranteed` if and only if all of the included containers define a resource request which is equal
+to the resource limit. The current implementation considers a pod to be `Burstable` if at least one container defines resource requests or limits. Any
+pods not matching the `Guaranteed` or `Burstable` QoS definitions are placed in the `BestEffort` QoS level.
+
+This proposal suggests modifying these definitions the following: 
+
+A pod is considered `Guaranteed` if either of the following two conditions apply: 
+1. All of the containers included in the pod define memory and CPU resource requests equal to resource limits
+1. The pod includes a pod-level memory and CPU resource request which is equal to the resource limit, even if one or more of the included containers doesn't define either a CPU or a memory request or limit.
+
+A pod is considered `Burstable` if at least one container has a resource request, or the pod itself has a resource request.
+
+A pod is considered `BestEffort` if there are not resource request or limits settings on any container or on the pod level itself.
+
+If a developer included a request and limit definition at the pod level, the developer's intent was for Kubernetes to attempt to guarantee resources 
+as a whole, for all the containers included in the pod. This should apply even if the developer is unable to preemptively split the resources correctly
+between all the containers contained in the pod. 
+
+Note that if a pod belongs to the `Guaranteed` QoS level as per the current definition, then defining a pod-level limit is redundant. If the pod level limit
```

Comment:
I think it would be best to keep as closely aligned as possible to the underlying Linux cgroup behavior for each resource. 

I will update the KEP.

Reply:
See above comment.  TL;DR We should define it explicitly or disallow it explicitly.  The worst choice is to not define it :)